crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -1043,6 +1043,80 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
 
         Ok(())
     }
+
+
+/// Represents a single line in a docstring.
+///
+/// This type is used to both represent the original lines in a docstring
+/// (the line will be borrowed) and also the newly formatted lines from code
+/// snippets (the line will be owned).
+#[derive(Clone, Debug)]
+struct DocstringLine<'src> {
+    /// The actual text of the line, not including the line terminator.
+    line: Cow<'src, str>,
+    /// The offset into the source document which this line corresponds to.
+    offset: TextSize,
+    /// Whether this is the last line in a docstring or not. "Last" lines have
+    /// some special treatment when printing.
+    is_last: bool,
+}
+
+impl<'src> DocstringLine<'src> {
+    /// Return this line, but with the given function applied to the text of
+    /// the line.
+    fn map(self, mut map: impl FnMut(&str) -> String) -> DocstringLine<'static> {
+        DocstringLine {
+            line: Cow::Owned(map(&self.line)),
+            ..self
+        }
+    }
+}
+/// The kind of code example observed in a docstring.
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum CodeExampleKind {
+    /// Code found in Python "doctests."
+    ///
+    /// Documentation describing doctests and how they're recognized can be
+    /// found as part of the Python standard library:
+    /// https://docs.python.org/3/library/doctest.html.
+    ///
+    /// (You'll likely need to read the regex matching used internally by the
+    /// doctest module to determine more precisely how it works.)
+    Doctest(CodeExampleDoctest),
+}
+
+/// State corresponding to a single doctest code example found in a docstring.
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct CodeExampleDoctest {
+    /// The indent observed in the first doctest line.
+    ///
+    /// More precisely, this corresponds to the whitespace observed before
+    /// the starting `>>> ` (the "PS1 prompt").
+    indent: String,
```

Comment:
Future perf improvement: Represent as an enum with `Tabs(count)`, `Spaces(count)`, and `Mixed(String)` to avoid allocating for each formatted line. 

Or could the indent be a ` &'a str` because it is the indent observed from the source text?

Reply:
I believe the indent could indeed be a `&'src str` here, but I went with the simpler approach because the extra allocation here seems marginal to me. We're already allocating other stuff, invoking the formatter and so on.

The reason why I chose to thread through a lifetime parameter for `DocstringLine` is that it gets created for every line in every docstring, regardless of whether there is a code snippet. So I wanted to be a little more cautious there. But by the time we create an `indent` string, we've already (very likely) committed ourselves to a bunch of extra costs too.